File: bindings/cpp/test/test_bindings.cpp
// SPDX-License-Identifier: GPL-3.0-only
// SPDX-FileCopyrightText: 2026 AnmiTaliDev <anmitalidev@nuros.org>

#include <apg++/apg.hpp>

#include <unistd.h>

#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <vector>

namespace
{

std::string make_tmp_dir(const char *prefix)
{
    std::string tmpl = std::string("/tmp/") + prefix + "-XXXXXX";
    std::vector<char> buf(tmpl.begin(), tmpl.end());
    buf.push_back('\0');
    if (!mkdtemp(buf.data()))
        throw apg::Error("mkdtemp failed");
    return std::string(buf.data());
}

void remove_tmp_db(const std::string &path)
{
    unlink((path + "/data.mdb").c_str());
    unlink((path + "/lock.mdb").c_str());
    rmdir(path.c_str());
}

apg::Package make_simple_package(const std::string &name,
                                  const std::string &version = "1.0")
{
    apg::Package pkg = apg::Package::create();
    struct package *raw = pkg.get();
    raw->meta->name = strdup(name.c_str());
    raw->meta->version = strdup(version.c_str());
    raw->pkg_path = strdup("/nonexistent/test.pkg");
    raw->installed_by_hand = true;
    return pkg;
}

void test_version()
{
    assert(apg::ver_compare("1.0", "2.0") < 0);
    assert(apg::ver_satisfies("2.0", VER_OP_GE, "1.0"));

    apg::DepConstraint c = apg::dep_constraint_parse("libfoo >= 2.0");
    assert(c.name == "libfoo");
    assert(c.op == VER_OP_GE);
    assert(c.version == "2.0");
    assert(apg::dep_constraint_to_str(c) == "libfoo >= 2.0");

    printf("test_version: PASS\n");
}

void test_database_roundtrip()
{
    std::string db_path = make_tmp_dir("apgxx-db");
    apg::Database db = apg::Database::open(db_path);

    apg::Package pkg = make_simple_package("cppxx-pkg");
    assert(db.add(pkg));

    auto fetched = db.get("cppxx-pkg");
    assert(fetched.has_value());
    assert(fetched->metadata().name() == "cppxx-pkg");
    assert(fetched->metadata().version() == "1.0");

    assert(db.set_hold("cppxx-pkg", true));

    apg::DbStats stats = db.stats();
    assert(stats.package_count >= 1);

    assert(db.remove("cppxx-pkg"));
    assert(!db.get("cppxx-pkg").has_value());

    remove_tmp_db(db_path);
    printf("test_database_roundtrip: PASS\n");
}

void test_dependency_graph()
{
    apg::PackageMetadata base = apg::PackageMetadata::create();
    base.get()->name = strdup("base");
    base.get()->version = strdup("1.0");

    apg::PackageMetadata app = apg::PackageMetadata::create();
    app.get()->name = strdup("app");
    app.get()->version = strdup("1.0");
    app.get()->dependencies.count = 1;
    app.get()->dependencies.items = static_cast<struct dep_constraint *>(
        malloc(sizeof(struct dep_constraint)));
    app.get()->dependencies.items[0] = dep_constraint_parse("base");

    apg::DependencyGraph graph;
    assert(graph.add(base) == DEP_OK);
    assert(graph.add(app) == DEP_OK);

    std::vector<std::string> order;
    assert(graph.resolve("app", order) == DEP_OK);
    assert(order.size() == 2);
    assert(order[0] == "base");
    assert(order[1] == "app");

    printf("test_dependency_graph: PASS\n");
}

void test_transaction_policy()
{
    std::string db_path = make_tmp_dir("apgxx-trans-db");
    apg::Database db = apg::Database::open(db_path);

    apg::Package pkg = make_simple_package("trans-pkg");
    apg::Transaction trans(db);
    assert(trans.add_install(pkg) == TRANS_OK);
    assert(trans.prepare() == TRANS_OK);

    apg::InstallPolicy policy;
    policy.require_signature = true;
    policy.keyring_dir = "/nonexistent/keys";
    trans.set_policy(policy);

    assert(trans.commit("/tmp") == TRANS_ERR_UNSIGNED);

    remove_tmp_db(db_path);
    printf("test_transaction_policy: PASS\n");
}

} // namespace

int main()
{
    test_version();
    test_database_roundtrip();
    test_dependency_graph();
    test_transaction_policy();
    printf("All C++ binding tests passed.\n");
    return 0;
}

File: bindings/python/tests/test_bindings.py
# SPDX-License-Identifier: GPL-3.0-only
# SPDX-FileCopyrightText: 2026 AnmiTaliDev <anmitalidev@nuros.org>

import apg


def make_pkg(name, version="1.0", provides=None):
    pkg = apg.Package.create()
    pkg.metadata.set_name(name)
    pkg.metadata.set_version(version)
    pkg.path = "/nonexistent/test.pkg"
    if provides:
        pkg.metadata.set_provides([provides])
    return pkg


def make_meta(name, version="1.0", provides=None):
    meta = apg.PackageMetadata.create()
    meta.set_name(name)
    meta.set_version(version)
    if provides:
        meta.set_provides([provides])
    return meta


def test_version():
    assert apg.ver_compare("1.0", "2.0") < 0
    assert apg.ver_satisfies("2.0", apg.VerOp.GE, "1.0")

    c = apg.dep_constraint_parse("libfoo >= 2.0")
    assert c.name == "libfoo"
    assert c.op == apg.VerOp.GE
    assert c.version == "2.0"
    assert apg.dep_constraint_to_str(c) == "libfoo >= 2.0"


def test_database_roundtrip(tmp_path):
    db = apg.Database.open(str(tmp_path))

    pkg = make_pkg("pathtest")
    pkg.path = "/var/cache/apg/pathtest-1.0.apg"
    assert db.add(pkg)

    fetched = db.get("pathtest")
    assert fetched is not None
    assert fetched.metadata.name == "pathtest"
    assert fetched.metadata.version == "1.0"
    assert fetched.path == "/var/cache/apg/pathtest-1.0.apg"

    assert db.set_hold("pathtest", True)

    pkg_count, _ = db.stats()
    assert pkg_count >= 1

    assert db.remove("pathtest")
    assert db.get("pathtest") is None

    fetched.close()
    pkg.close()
    db.close()


def test_db_list_and_search(tmp_path):
    db = apg.Database.open(str(tmp_path))

    pkg = make_pkg("searchable-pkg")
    pkg.metadata.set_description("a package used for search tests")
    assert db.add(pkg)

    names = [p.metadata.name for p in db.list()]
    assert "searchable-pkg" in names

    results = db.search("searchable")
    assert any(p.metadata.name == "searchable-pkg" for p in results)

    pkg.close()
    db.close()


def test_dependency_graph_prefers_installed():
    nginx = make_meta("nginx", "1.0", "webserver")
    apache = make_meta("apache", "1.0", "webserver")
    app = make_meta("app")
    app.set_dependencies([apg.dep_constraint_parse("webserver")])

    g = apg.DependencyGraph()
    g.add(apache)
    g.add(app)
    g.add_installed(nginx)

    err, order = g.resolve("app")
    assert err == 0
    assert order == ["nginx", "app"]

    dot = g.export_dot()
    assert "digraph libapg_deps" in dot
    assert '"nginx" [style=filled, fillcolor="lightgreen"]' in dot

    g.close()
    nginx.close()
    apache.close()
    app.close()


def test_dependency_graph_falls_back_without_installed():
    first = make_meta("apache", "1.0", "webserver")
    second = make_meta("caddy", "1.0", "webserver")
    app = make_meta("app")
    app.set_dependencies([apg.dep_constraint_parse("webserver")])

    g = apg.DependencyGraph()
    g.add(first)
    g.add(second)
    g.add(app)

    err, order = g.resolve("app")
    assert err == 0
    assert order == ["apache", "app"]

    g.close()
    first.close()
    second.close()
    app.close()


def test_transaction_policy(tmp_path):
    db = apg.Database.open(str(tmp_path))
    pkg = make_pkg("trans-pkg")

    trans = apg.Transaction(db)
    assert trans.add_install(pkg) == apg.TransError.OK
    assert trans.prepare() == apg.TransError.OK

    policy = apg.InstallPolicy(
        require_signature=True, keyring_dir="/nonexistent/keys"
    )
    trans.set_policy(policy)

    assert trans.commit("/tmp") == apg.TransError.UNSIGNED

    trans.close()
    pkg.close()
    db.close()


def test_transaction_prefer_provider_overrides_default(tmp_path):
    db = apg.Database.open(str(tmp_path))

    nginx_installed = make_pkg("nginx", "1.0", "webserver")
    assert db.add(nginx_installed)

    caddy = make_pkg("caddy", "2.0", "webserver")
    app = make_pkg("app")
    app.metadata.set_dependencies(
        [apg.dep_constraint_parse("webserver >= 2.0")]
    )

    trans = apg.Transaction(db)
    assert trans.add_install(caddy) == apg.TransError.OK
    assert trans.add_install(app) == apg.TransError.OK
    trans.prefer_provider("webserver", "caddy")

    assert trans.prepare() == apg.TransError.OK
    plan = trans.plan
    assert [step.pkg_name for step in plan] == ["caddy", "app"]

    trans.close()
    caddy.close()
    app.close()
    nginx_installed.close()
    db.close()


def test_transaction_default_prefers_installed_provider(tmp_path):
    db = apg.Database.open(str(tmp_path))

    nginx_installed = make_pkg("nginx", "1.0", "webserver")
    assert db.add(nginx_installed)

    caddy = make_pkg("caddy", "2.0", "webserver")
    app = make_pkg("app")
    app.metadata.set_dependencies(
        [apg.dep_constraint_parse("webserver >= 2.0")]
    )

    trans = apg.Transaction(db)
    assert trans.add_install(caddy) == apg.TransError.OK
    assert trans.add_install(app) == apg.TransError.OK

    assert trans.prepare() != apg.TransError.OK

    trans.close()
    caddy.close()
    app.close()
    nginx_installed.close()
    db.close()


def test_audit_log(tmp_path):
    db = apg.Database.open(str(tmp_path))
    pkg = make_pkg("audit-pkg")

    trans = apg.Transaction(db)
    trans.add_install(pkg)
    trans.prepare()
    trans.set_policy(apg.InstallPolicy(require_signature=False))
    trans.commit("/tmp")

    entries = apg.audit.read_all(db)
    assert len(entries) >= 1
    last = entries[-1]
    assert last.op == apg.JournalOp.INSTALL
    assert last.pkg_name == "audit-pkg"

    for e in entries:
        e.close()
    trans.close()
    pkg.close()
    db.close()
